### ISSUE TITLE
Add directional arrow heads to edges

### ISSUE THREAD

**Issue by jdanish** [netcreateorg/netcreate-2018#262]
If it is relatively quick, it'd be nice to display directionality on edges (small arrowheads in each direction that an edge includes).

Short-term, adding the arrow-heads would be nice on any edge. If A-> B and B -> A it would appear bi-directional.
Long-term it'd be nice to specify an edge as bi-directional using a check-box in the edge.
It'd be nice if there was an option in the template to turn arrowhead display on or off. This can also wait (default on)

**Comment by jdanish** [comment]
Clarification: these arrows are intended to be in the graph. I am assuming that for now, source and target convey direction. This means that short-term, anything that is bi-directional will be confusing, and long-term we can figure that out (likely a checkbox).

**Comment by benloh** [comment]
There are also two complicating factors:
* How to handle colored edges, especially if there are two different colors involved.
* How to handle thick edges -- show the arrow be larger too?  Is it still visible small?

**Comment by benloh** [comment]
It looks like it's relatively straightforward to implement on the side if are not worried about thickness (and possibly colors).  Here are a few resources/examples:
* https://stackoverflow.com/questions/36965610/how-to-add-an-arrow-links-in-d3-js
* https://gist.github.com/fancellu/2c782394602a93921faff74e594d1bb1
* https://itecnote.com/tecnote/r-arrows-on-links-in-d3js-force-layout/
* https://www.appsloveworld.com/d3js/100/1/display-an-arrow-head-in-the-middle-of-d3-force-layout-link

**Comment by jdanish** [comment]
For handling thick edges I think let's see what happens and assume we may need to scale up to match thicker edges?

For color, here is what CHAT GPT says are the options with a comment after each:

- Overlay Colors: One common practice is to overlay the colors of the different edge types. For example, if there are two edge types connecting two nodes, one colored blue and the other red, the visualization tool might display the overlapping region with a mix of the two colors, creating a purple hue. **(OK)**
- Weighted Colors: Another approach is to assign different weights or opacities to the edge types. The tool may use a blending technique to combine the colors based on their respective weights, creating a visual representation that reflects the relative importance or strength of each edge type. **(seems too much)**
- Sequential Display: Some visualization tools choose to display the edges sequentially rather than overlapping them. In this case, the tool may iterate through the different edge types, displaying each one individually. This allows users to observe the edges separately but may require additional interaction or animation to view all the connections.  **(seems too much)**
- Parallel Edges: Alternatively, the visualization tool may choose to display the different edge types as parallel lines. For instance, if there are two edge types, the tool may render one edge above the other, preserving their distinct colors and avoiding visual blending. This approach can be useful when it's important to maintain the clarity of individual edge types. **would be cool but I presume is challenging**
What happens now?

**Comment by benloh** [comment]
@jdanish Here's an interesting idea.  What do you think of putting the arrow in the middle?  That would solve some node size issues, though it doesn't solve edge size issues...
![image](https://github.com/netcreateorg/netcreate-2018/assets/1403057/aea75d38-0313-49a4-b77b-a4f9a2eb6dc6)

**Comment by kalanicraig** [comment]
Can we use parallel straight edges, or arrows facing away from each other
on a single straight edge with some padding between the “bottom” of the
arrowhead so that the directionality is visible, rather than curved edges.
With many curved edges, following each edge in a crowded part of the viz
gets complicated.

**Comment by benloh** [comment]
Sorry I wasn't suggesting curved edges but straight edges with the arrows in the center of the line segment instead of at the ends of the lines.  But I can see why that might be a problem with long lines.  So forget I suggested this!

**Comment by kalanicraig** [comment]
Well, maybe not forget. One possibility is to show
oppositional/antagonistic edges with the arrows in the middle pointing
toward each other, but that’s a question for later. In the meantime, could
we use a-index to put a small contrast-color arrow over the end of edges
that are displaying with a very wide edge weight so that the directionality
is still visible? Dunno if that’s too much of a pain.

**Comment by jdanish** [comment]
Ok thanks! ----(from my iPhone)Joshua Danishhttp://www.joshuadanish.comOn Jun 7, 2023, at 4:14 PM, benloh ***@***.***> wrote:﻿
I'm still trying to figure out how the arrows work between d3 and SVG, but I'll see what I can do.  One thing that might not be easy is coloring the arrows.  So we might have to live with black arrows.

**Comment by benloh** [comment]
@jdanish @kalanicraig It looks like for arrowheads we're going to have to do a lot of case-by-case handling of different situations in order to make this work.  Because SVG provides a `<marker>` object for rendering arrows, I had thought this would be relatively easy to implement, but markers are fairly limited in what they can do out of the box.  We'll have to build a lot of workarounds.  Some issues:

* Links are drawn to the center of nodes, so large node sizes will require us to adjust where the arrowhead is drawn, otherwise the node will cover the arrowhead.
* Arrowheads will have to be one color -- they won't match the color of their edge.  If this is absolutely critical it is possible to do, but basically we'd be creating two arrowheads for each edge color.
* Currently we are blindly drawing every defined edge.  So if there is more than one edge between two nodes, we're drawing multiple edges.  The problem with this is the subsequently drawn edges will cover the previously drawn arrowheads.  And there is no easy way to set a z-index in SVG.  So we probably need to add a pre-render pass where we simplify the edges, stripping out duplicates -- this is probably more efficient anyway.
* Arrowheads on thin strokes can not be seen at all.  By default arrowheads render relative to the size of the stroke.  This works well for large strokes (e.g. > 3), but for small strokes the arrowheads are essentially invisible.  In order to get around this we probably need to use a fixed size technique for small strokes and a second relative-to-stroke-size technique for larger arrowheads.
* The larger arrowheads can get REALLY huge though, taking up the whole line.  So we probably want to add a third technique that tries to limit the length of the arrowhead.

Do you have any sense of what max edge weight you would like to set?  20?  50?  That'll help me figure out how to render the arrow.

Updating estimate to 20.

**Comment by benloh** [comment]
Makes sense.  Though if this is something you definitely want to do for this round of funding, we should just go ahead and do it since I'm already in the middle of it.  This won't prevent us from moving forward on the other infrastructure stuff.

**Comment by jdanish** [comment]
We are of mixed opinions and would rather prioritize other things if it comes to it. So let's hold off for now, thanks!
